<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Reference for rnaseqflow &mdash; rnaseqflow 0.2.0 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="rnaseqflow 0.2.0 documentation" href="index.html" />
    <link rel="prev" title="What Is It?" href="README.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="reference-for-rnaseqflow">
<h1>Reference for rnaseqflow<a class="headerlink" href="#reference-for-rnaseqflow" title="Permalink to this headline">¶</a></h1>
<p>rnaseqflow can do things</p>
<div class="section" id="module-rnaseqflow.__main__">
<span id="rnaseqflow-main"></span><h2><code class="docutils literal"><span class="pre">rnaseqflow.__main__</span></code><a class="headerlink" href="#module-rnaseqflow.__main__" title="Permalink to this headline">¶</a></h2>
<p>Created on Dec 13, 2015</p>
<p>&#64;author: justinpalpant</p>
<p>This file is part of the Jarvis Lab RNAseq Workflow program.</p>
<p>RNAseq Workflow is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.</p>
<p>RNAseq Workflow is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License along with
RNAseq Workflow. If not, see <a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>.</p>
<dl class="function">
<dt id="rnaseqflow.__main__.main">
<code class="descclassname">rnaseqflow.__main__.</code><code class="descname">main</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.__main__.main" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is installed as a console script entry point by setuptools</p>
<p>It uses the command line arguments specified by opts() to generate a
Workflow object and adds to it several WorkflowStages.</p>
<p>If the needed command line arguments are not passed, the user is asked to
enter them.</p>
<p>The generated Workflow object is then executed with run()</p>
</dd></dl>

<dl class="function">
<dt id="rnaseqflow.__main__.opts">
<code class="descclassname">rnaseqflow.__main__.</code><code class="descname">opts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.__main__.opts" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-rnaseqflow.workflow">
<span id="rnaseqflow-workflow"></span><h2><code class="docutils literal"><span class="pre">rnaseqflow.workflow</span></code><a class="headerlink" href="#module-rnaseqflow.workflow" title="Permalink to this headline">¶</a></h2>
<p>Created on Dec 12, 2015</p>
<p>&#64;author: justinpalpant</p>
<p>Copyright 2015 Justin Palpant</p>
<p>This file is part of the Jarvis Lab RNAseq Workflow program.</p>
<p>RNAseq Workflow is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.</p>
<p>RNAseq Workflow is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License along with
RNAseq Workflow. If not, see <a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>.</p>
<dl class="class">
<dt id="rnaseqflow.workflow.FastQMCFTrimPairs">
<em class="property">class </em><code class="descclassname">rnaseqflow.workflow.</code><code class="descname">FastQMCFTrimPairs</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimPairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Trim adapter sequences from files using fastq-mcf in paired-end mode</p>
<dl class="docutils">
<dt>Input:</dt>
<dd>A flat set of files to be passed into fastq-mcf in pairs</dd>
<dt>Output:</dt>
<dd>A flat set of trimmed file names</dd>
<dt>Args used:</dt>
<dd>&#8211;root: the folder where trimmed files will be placed
&#8211;adapters: the filepath of the fasta adapters file
&#8211;fastq: the location of the fastq-mcf executable
&#8211;fastq_args: a string of arguments to pass directly to fastq-mcf
&#8211;quiet: silence fastq-mcf&#8217;s output if given</dd>
</dl>
<dl class="attribute">
<dt id="rnaseqflow.workflow.FastQMCFTrimPairs.logger">
<code class="descname">logger</code><em class="property"> = &lt;logging.Logger object&gt;</em><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimPairs.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="rnaseqflow.workflow.FastQMCFTrimPairs.run">
<code class="descname">run</code><span class="sig-paren">(</span><em>stage_input</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimPairs.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Trim files one at a time using fastq-mcf</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>stage_input - a flat list of file names</dd>
<dt>Returns:</dt>
<dd>A flat list of trimmed file names</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="rnaseqflow.workflow.FastQMCFTrimPairs.spec">
<code class="descname">spec</code><em class="property"> = '3.1'</em><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimPairs.spec" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="rnaseqflow.workflow.FastQMCFTrimSolo">
<em class="property">class </em><code class="descclassname">rnaseqflow.workflow.</code><code class="descname">FastQMCFTrimSolo</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimSolo" title="Permalink to this definition">¶</a></dt>
<dd><p>Trim adapter sequences from files using fastq-mcf one file at a time</p>
<dl class="docutils">
<dt>Input:</dt>
<dd>A flat set of files to be passed into fastq-mcf file-by-file</dd>
<dt>Output:</dt>
<dd>A flat set of trimmed file names</dd>
<dt>Args used:</dt>
<dd>&#8211;root: the folder where trimmed files will be placed
&#8211;adapters: the filepath of the fasta adapters file
&#8211;fastq: the location of the fastq-mcf executable
&#8211;fastq_args: a string of arguments to pass directly to fastq-mcf
&#8211;quiet: silence fastq-mcf&#8217;s output if given</dd>
</dl>
<dl class="attribute">
<dt id="rnaseqflow.workflow.FastQMCFTrimSolo.logger">
<code class="descname">logger</code><em class="property"> = &lt;logging.Logger object&gt;</em><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimSolo.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="rnaseqflow.workflow.FastQMCFTrimSolo.run">
<code class="descname">run</code><span class="sig-paren">(</span><em>stage_input</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimSolo.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Trim files one at a time using fastq-mcf</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>stage_input - a flat list of file names</dd>
<dt>Returns:</dt>
<dd>A flat list of trimmed file names</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="rnaseqflow.workflow.FastQMCFTrimSolo.spec">
<code class="descname">spec</code><em class="property"> = '3.0'</em><a class="headerlink" href="#rnaseqflow.workflow.FastQMCFTrimSolo.spec" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="rnaseqflow.workflow.FindFiles">
<em class="property">class </em><code class="descclassname">rnaseqflow.workflow.</code><code class="descname">FindFiles</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.FindFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Find files recursively in a folder</p>
<dl class="docutils">
<dt>Input:</dt>
<dd>No input is required for this WorkflowStage</dd>
<dt>Output:</dt>
<dd>A flat set of file path strings</dd>
<dt>Args used:</dt>
<dd>&#8211;root: the folder in which to start the search
&#8211;ext: the file extention to search for</dd>
</dl>
<dl class="attribute">
<dt id="rnaseqflow.workflow.FindFiles.logger">
<code class="descname">logger</code><em class="property"> = &lt;logging.Logger object&gt;</em><a class="headerlink" href="#rnaseqflow.workflow.FindFiles.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="rnaseqflow.workflow.FindFiles.run">
<code class="descname">run</code><span class="sig-paren">(</span><em>stage_input</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.FindFiles.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the recursive file finding stage</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>stage_input - not used, only for the interface</dd>
<dt>Returns:</dt>
<dd>A flat list of files found with the correct extension</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="rnaseqflow.workflow.FindFiles.spec">
<code class="descname">spec</code><em class="property"> = '1'</em><a class="headerlink" href="#rnaseqflow.workflow.FindFiles.spec" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="rnaseqflow.workflow.MergeSplitFiles">
<em class="property">class </em><code class="descclassname">rnaseqflow.workflow.</code><code class="descname">MergeSplitFiles</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.MergeSplitFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge files by the identifying sequence and direction</p>
<dl class="docutils">
<dt>Input:</dt>
<dd>An iterable of file names to be grouped and merged</dd>
<dt>Output:</dt>
<dd>A flat set of merged filenames</dd>
<dt>Args used:</dt>
<dd>&#8211;root: the folder where merged files will be placed
&#8211;ext: the file extention to be used for the output files
&#8211;blocksize: number of kilobytes to use as a copy block size</dd>
</dl>
<dl class="attribute">
<dt id="rnaseqflow.workflow.MergeSplitFiles.logger">
<code class="descname">logger</code><em class="property"> = &lt;logging.Logger object&gt;</em><a class="headerlink" href="#rnaseqflow.workflow.MergeSplitFiles.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="rnaseqflow.workflow.MergeSplitFiles.run">
<code class="descname">run</code><span class="sig-paren">(</span><em>stage_input</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.MergeSplitFiles.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the merge files operation</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>stage_input - a list of files to be organized and merged</dd>
<dt>Returns:</dt>
<dd>A flat list of merged files</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="rnaseqflow.workflow.MergeSplitFiles.spec">
<code class="descname">spec</code><em class="property"> = '2'</em><a class="headerlink" href="#rnaseqflow.workflow.MergeSplitFiles.spec" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="rnaseqflow.workflow.Workflow">
<em class="property">class </em><code class="descclassname">rnaseqflow.workflow.</code><code class="descname">Workflow</code><a class="headerlink" href="#rnaseqflow.workflow.Workflow" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute a simple series of steps used to preprocess RNAseq files</p>
<dl class="method">
<dt id="rnaseqflow.workflow.Workflow.append">
<code class="descname">append</code><span class="sig-paren">(</span><em>item</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.Workflow.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a WorkflowStage to the workflow</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>item - a WorkflowStage to execute</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="rnaseqflow.workflow.Workflow.insert">
<code class="descname">insert</code><span class="sig-paren">(</span><em>idx</em>, <em>item</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.Workflow.insert" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert a WorkflowStage into the workflow</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>item - a WorkflowStage to execute</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="rnaseqflow.workflow.Workflow.logger">
<code class="descname">logger</code><em class="property"> = &lt;logging.Logger object&gt;</em><a class="headerlink" href="#rnaseqflow.workflow.Workflow.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="rnaseqflow.workflow.Workflow.run">
<code class="descname">run</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.Workflow.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Allows the user to select a directory and processes all files within
that directory</p>
<p>This function is the primary function of the Workflow class.  All other
functions are written as support for this function, at the moment</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="rnaseqflow.workflow.WorkflowStage">
<em class="property">class </em><code class="descclassname">rnaseqflow.workflow.</code><code class="descname">WorkflowStage</code><a class="headerlink" href="#rnaseqflow.workflow.WorkflowStage" title="Permalink to this definition">¶</a></dt>
<dd><p>Interfaces for a stage of a Workflow</p>
<p>Subclasses must override the run method, which takes and verifies arbitrary
input, processes it, and returns some output</p>
<dl class="attribute">
<dt id="rnaseqflow.workflow.WorkflowStage.logger">
<code class="descname">logger</code><em class="property"> = &lt;logging.Logger object&gt;</em><a class="headerlink" href="#rnaseqflow.workflow.WorkflowStage.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="classmethod">
<dt id="rnaseqflow.workflow.WorkflowStage.longhelp">
<em class="property">classmethod </em><code class="descname">longhelp</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.WorkflowStage.longhelp" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="rnaseqflow.workflow.WorkflowStage.run">
<code class="descname">run</code><span class="sig-paren">(</span><em>stage_input</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.WorkflowStage.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Attempt to process the provided input according to the rules of the
subclass</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd><dl class="first last docutils">
<dt>stage_input - an arbitrary input to be processed, usually a list of</dt>
<dd>file names or file-like objects.  The subclass must typecheck
the input as necessary, and define what input it takes</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>output - the results of the processing of this workflow item</dd>
</dl>
</dd></dl>

<dl class="classmethod">
<dt id="rnaseqflow.workflow.WorkflowStage.shorthelp">
<em class="property">classmethod </em><code class="descname">shorthelp</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.workflow.WorkflowStage.shorthelp" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="rnaseqflow.workflow.WorkflowStage.spec">
<code class="descname">spec</code><a class="headerlink" href="#rnaseqflow.workflow.WorkflowStage.spec" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract class property, override with &#64;classmethod</p>
<p>Used by the help method to specify available WorkflowItems</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-rnaseqflow.cliutils">
<span id="rnaseqflow-cliutils"></span><h2><code class="docutils literal"><span class="pre">rnaseqflow.cliutils</span></code><a class="headerlink" href="#module-rnaseqflow.cliutils" title="Permalink to this headline">¶</a></h2>
<p>Created on Mar 15, 2016</p>
<p>&#64;author: justinpalpant</p>
<p>Copyright 2016 Justin Palpant</p>
<p>This file is part of the Jarvis Lab RNAseq Workflow program.</p>
<p>RNAseq Workflow is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.</p>
<p>RNAseq Workflow is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License along with
RNAseq Workflow. If not, see <a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>.</p>
<dl class="class">
<dt id="rnaseqflow.cliutils.ArgFiller">
<em class="property">class </em><code class="descclassname">rnaseqflow.cliutils.</code><code class="descname">ArgFiller</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.cliutils.ArgFiller" title="Permalink to this definition">¶</a></dt>
<dd><p>An interactive method of filling in arguments not given at runtime</p>
<p>Code completion taken from <a class="reference external" href="https://gist.github.com/iamatypeofwalrus/5637895">https://gist.github.com/iamatypeofwalrus/5637895</a></p>
<dl class="method">
<dt id="rnaseqflow.cliutils.ArgFiller.fill">
<code class="descname">fill</code><span class="sig-paren">(</span><em>args_needed</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.cliutils.ArgFiller.fill" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the needed arguments to self.args if they are not there</p>
<p>Asks the user for input for each of the arguments</p>
</dd></dl>

<dl class="attribute">
<dt id="rnaseqflow.cliutils.ArgFiller.logger">
<code class="descname">logger</code><em class="property"> = &lt;logging.Logger object&gt;</em><a class="headerlink" href="#rnaseqflow.cliutils.ArgFiller.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="rnaseqflow.cliutils.ArgFiller.pathCompleter">
<code class="descname">pathCompleter</code><span class="sig-paren">(</span><em>text</em>, <em>state</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.cliutils.ArgFiller.pathCompleter" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the tab completer for systems paths.
Only tested on <a href="#id1"><span class="problematic" id="id2">*</span></a>nix systems</p>
</dd></dl>

<dl class="method">
<dt id="rnaseqflow.cliutils.ArgFiller.set_path_complete">
<code class="descname">set_path_complete</code><span class="sig-paren">(</span><em>enable</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.cliutils.ArgFiller.set_path_complete" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="rnaseqflow.cliutils.all_subclasses">
<code class="descclassname">rnaseqflow.cliutils.</code><code class="descname">all_subclasses</code><span class="sig-paren">(</span><em>cls</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.cliutils.all_subclasses" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="rnaseqflow.cliutils.firstline">
<code class="descclassname">rnaseqflow.cliutils.</code><code class="descname">firstline</code><span class="sig-paren">(</span><em>docstring</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.cliutils.firstline" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="rnaseqflow.cliutils.trim">
<code class="descclassname">rnaseqflow.cliutils.</code><code class="descname">trim</code><span class="sig-paren">(</span><em>docstring</em><span class="sig-paren">)</span><a class="headerlink" href="#rnaseqflow.cliutils.trim" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Reference for rnaseqflow</a><ul>
<li><a class="reference internal" href="#module-rnaseqflow.__main__"><code class="docutils literal"><span class="pre">rnaseqflow.__main__</span></code></a></li>
<li><a class="reference internal" href="#module-rnaseqflow.workflow"><code class="docutils literal"><span class="pre">rnaseqflow.workflow</span></code></a></li>
<li><a class="reference internal" href="#module-rnaseqflow.cliutils"><code class="docutils literal"><span class="pre">rnaseqflow.cliutils</span></code></a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="README.html" title="previous chapter">What Is It?</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/ref_rnaseqflow.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Justin Palpant.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
      |
      <a href="_sources/ref_rnaseqflow.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>